### DISCUSSION TITLE
Modifying stable-diffusion demo for SDXL inpainting

### DISCUSSION THREAD

**Discussion by mneilly** [huggingface/candle#2350] [Q&A]
I'm trying to modify the stable diffusion demo to add SDXL inpainting [here](https://github.com/huggingface/candle/compare/main...mneilly:candle:mneilly/sdxl-inpainting). 

I have added options to load the mask image, generate the mask and masked image latents and feed them into the unet. I can compile, run and generate an image but the masked region doesn't contain an image based on the prompt and the colors for the unmasked region contain a color distorted version of the original. I've been comparing against the pytorch model but I seem to be overlooking something...

The final latents look like this before replacing the unmasked region with the original pixels:

![latents-small](https://github.com/user-attachments/assets/69f90c8d-5550-4c0b-9966-5995038d2936)

and after replacing them:

![sd_final-small](https://github.com/user-attachments/assets/5d69a16e-4d21-4785-bd4e-56e007948415)

I'm using the model from: https://huggingface.co/diffusers/stable-diffusion-xl-1.0-inpainting-0.1 and running as follows:

```
cargo run
--example stable-diffusion
--release
--features cuda,cudnn,flash-attn
--
--start-img images/view.png
--mask-image images/view-mask.png
--prompt "flying saucer in the sky"
--sd-version "xl-inpainting"
--use-f16
--use-flash-attn
--unet-weights  diffusers/stable-diffusion-xl-1.0-inpainting-0.1/unet/diffusion_pytorch_model.fp16.safetensors
--vae-weights   diffusers/stable-diffusion-xl-1.0-inpainting-0.1/vae/diffusion_pytorch_model.fp16.safetensors
--clip-weights  diffusers/stable-diffusion-xl-1.0-inpainting-0.1/text_encoder/model.fp16.safetensors
--clip2-weights diffusers/stable-diffusion-xl-1.0-inpainting-0.1/text_encoder_2/model.fp16.safetensors
```

I'll keep debugging but I'm wondering if anybody has any insight into what I might be missing.

**Comment by mneilly** [comment]
Alright, seems it might be a parameter issue and not necessarily bad code so now I can play around with it.

![sd_final_small](https://github.com/user-attachments/assets/e2e64fd1-c9e5-496c-83b8-1fa6176bfe84)

**Comment by AlpineVibrations** [comment]
any luck?

**Comment by mneilly** [comment]
I stopped looking at it for quite a while but recently wanted it working again. It isn't perfect but getting there.

```
cargo run 
--example stable-diffusion 
--features cuda,cudnn,flash-attn 
-- 
--prompt "a cute baby lion sitting on a park bench" 
--sd-version xl-inpaint 
--use-f16 
--use-flash-attn 
--img2img overture-creations-5sI6fQgYIuo_big.png 
--mask-path overture-creations-5sI6fQgYIuo_mask_big.png 
--guidance-scale 7 
--num-samples 10
```

![montage](https://github.com/user-attachments/assets/0e88800b-a229-455d-82d3-6b0cf6c8c5ef)

**Reply by john09282922** [reply]
Hi, can you tell me how to make better?

**Comment by AlpineVibrations** [comment]
i ran it a few times and not getting any good results yet.
